freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/RepairInstancesService.java

```diff
@@ -255,4 +273,37 @@ public OperationStatus rebootInstances(String accountId, RebootInstancesRequest
         }
         return operationToOperationStatusConverter.convert(operation);
     }
+
+    public DescribeFreeIpaResponse rebuild(String accountId, RebuildRequest rebuildRequest) {
+        Stack stack = stackService.findByCrnAndAccountIdWithListsEvenIfTerminated(rebuildRequest.getEnvironmentCrn(), accountId, rebuildRequest.getSourceCrn())
+                .orElse(null);
+        if (stack == null) {
+            String error = String.format("Stack not found with CRN %s in environment %s.", rebuildRequest.getSourceCrn(), rebuildRequest.getEnvironmentCrn());
+            LOGGER.error(error);
+            throw new BadRequestException(error);
+        }
+        Optional<Stack> nonTerminatedStack = stackService.findByEnvironmentCrnAndAccountId(rebuildRequest.getEnvironmentCrn(), accountId);
+        if (nonTerminatedStack.isPresent()) {
+            String error = "There is a stack which hasn't been terminated.";
+            LOGGER.error(error);
+            throw new BadRequestException(error);
+        }
+        renameStackIfNeeded(stack);
+        CreateFreeIpaRequest createFreeIpaRequest = stackToCreateFreeIpaRequestConverter.convert(stack);
+
+        return freeIpaCreationService.launchFreeIpa(createFreeIpaRequest, accountId);
+    }
+
+    void renameStackIfNeeded(Stack stack) {
+        if (!stack.isDeleteCompleted()) {
+            throw new BadRequestException(String.format("The stack %s has not been terminated", stack.getResourceCrn()));
+        }
+        Long teriminated = stack.getTerminated();
+        String originalName = stack.getName();
+        String deletionTime = StringUtils.substringAfterLast(originalName, DELETED_NAME_DELIMETER);
+        if (teriminated == -1 || !deletionTime.equals(teriminated.toString())) {
+            terminationService.finalizeTermination(stack.getId());
+            LOGGER.info("Updating terminated stack name from {}", originalName);
```

Comment:
could you move this line up to before the actual call? also please add `terminated` timestamp to the logmessage, it might be helpful if it doesn't match.
also `teriminated` has an extra `i` :)

Reply:
done